core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+   *   reference queue that will be used to track garbage collected elements
+   * @param nextRef
+   *   next reference node in the chain
+   * @tparam V
+   *   type of the element
+   */
+  protected class Ref[V](
+    val hash: Int,
+    element: V,
+    queue: ReferenceQueue[V],
+    var nextRef: Ref[V]
+  ) extends WeakReference[V](element, queue)
+
+  /**
+   * Access options for the update operation.
+   */
+  protected object UpdateOptions extends Enumeration {
+    type UpdateOption = Value
+    val RestructureBefore, RestructureAfter, SkipIfEmpty, Resize = Value
+  }
+
+  /**
+   * Supported results of the update operation.
+   */
+  protected object UpdateResults extends Enumeration {
+    type UpdateResult = Value
+    val None, AddElement, RemoveElement = Value
+  }
+
+  /**
+   * Base class for all update operations.
+   */
+  protected abstract class UpdateOperation[V](val options: UpdateOptions.UpdateOption*) {
+    def execute(oldRef: Ref[V]): UpdateResults.UpdateResult
+  }
+
+  /**
+   * Calculates the shift for configuration parameters. The shift is used to
+   * calculate the size of the segments and the size of the reference array.
+   */
+  private[internal] def calculateShift(minimumValue: Int, maximumValue: Int): Int = {
+    var shift = 0
+    var value = 1
+    while (value < minimumValue && value < maximumValue) {
+      value <<= 1
+      shift += 1
+    }
+    shift
+  }
+
+}
+
+/**
+ * A `Set` data type that uses weak references to store its elements in highly
+ * concurrent environment. This is faster alternative to
+ * `Collections.synchronizedSet(Collections.newSetFromMap(new WeakHashMapA,
+ * java.lang.Boolean))`.
+ *
+ * Implementation of this class is a combination of some existing solutions:
+ *   - `ConcurrentHashMap` from JDK 7 - for concurrency control using segments
+ *   - `WeakHashMap` - for storing weak references to elements
+ *   - `ConcurrentReferenceHashMap` from Spring Framework - the reference map
+ *     that combines references & with concurrent segments
+ *
+ * This implementation offers better performance than synchronized set for
+ * concurrent operations, but might be a little bit slower for serial
+ * (single-threaded only) access.
+ *
+ * @param initialCapacity
+ *   initial size of the set (default
+ *   [[ConcurrentWeakHashSet.DefaultInitialCapacity]])
+ * @param loadFactor
+ *   load factor for the set that defines how fast it will grow (default
+ *   [[ConcurrentWeakHashSet.DefaultLoadFactor]])
+ * @param concurrencyLevel
+ *   approximate number of threads that will access the set concurrently
+ *   (default [[ConcurrentWeakHashSet.DefaultConcurrencyLevel]])
+ * @tparam V
+ *   type of the elements stored in the set
+ */
+private[zio] class ConcurrentWeakHashSet[V](
+  initialCapacity: Int = ConcurrentWeakHashSet.DefaultInitialCapacity,
+  loadFactor: Float = ConcurrentWeakHashSet.DefaultLoadFactor,
+  concurrencyLevel: Int = ConcurrentWeakHashSet.DefaultConcurrencyLevel
```

Comment:
Currently the `loadFactor` and `concurrencyLevel` will require 4 bytes each, increasing storage by a total of 8 bytes, and contributing to heap churn. I assume, but do not know yet (I am reviewing), that these fields will be access across the `ConcurrentWeakHashSet`, rather than just inside initialization, in which case Scala will generate fields for them and store them inside every instance of `ConcurrentWeakHashSet`.

Consider making these `def` and giving them default. The user can extend the class and override the methods if they wish to change it. 

For similar reasons, be sure not to capture over `initialCapacity` in any other place than initialization of the fields of the class, or else, Scala will store a reference to `initialCapacity` inside the class, which will increase storage requirements by 4 bytes.